### DISCUSSION TITLE
Events and arrays (mqtt)

### DISCUSSION THREAD

**Discussion by CodingEJ** [arendst/Tasmota#12472] [Q&A] [answered]
Hello,
I have below MQTT data, where I want to switch on/off relays on basis of the data in zone_list. 

topic: "SIP/zones"
payload: "{"zone_list": [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0], "zone_dict": {"S01": 0, "S02": 0, "S03": 0, "S04": 0, "S05": 0, "S06": 0, "S07": 0, "S08": 0, "S09": 0, "S10": 0, "S11": 0, "S12": 0, "S13": 0, "S14": 0, "S15": 0, "S16": 0}, "master_on": 0}"

I have compiled tasmota 9.5.0 to include mqtt subscribe and was expecting the following rule to work:

Rule1 
ON mqtt#connected DO Subscribe SIPupdate, SIP/zones, zone_list ENDON
ON Event#SIPupdate[1] DO Power1 %value% ENDON

However, the result is that no value if passed.
RUL: EVENT#SIPUPDATE[1] performs "Power1 "

Any pointer on what I am doing wrong would be appreciated.

regards,
EJ

**Comment by barbudor** [comment, marked as answer]
I see nowhere that arrays are supported in triggers
However assuming that `zone_list[1]` is the same as `zone_dict["S02"]`, then below should do what you are looking for:
```
Rule1
ON mqtt#connected DO Subscribe SIPupdate, SIP/zones, zone_dict ENDON
ON Event#SIPupdate#S02 DO Power1 %value% ENDON
```
Not tested though

**Reply by CodingEJ** [reply]
In the rule documentation (https://tasmota.github.io/docs/Rules/) there is some piece that talks about triggers and arrays:

<html>
<body>
<p style="box-sizing: inherit; color: rgba(0, 0, 0, 0.87); font-family: Barlow, -apple-system, BlinkMacSystemFont, Helvetica, Arial, sans-serif; font-size: 17.6px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(255, 255, 255); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">When the payload consists of an array of data eg:<span> </span><code style="box-sizing: inherit; color: var(--md-code-fg-color); font-feature-settings: &quot;kern&quot;; font-family: var(--md-code-font-family,_),SFMono-Regular,Consolas,Menlo,monospace; direction: ltr; background-color: var(--md-code-bg-color); border-radius: 0.1rem; -webkit-box-decoration-break: clone; font-size: 0.85em; padding: 0px 0.294118em; word-break: break-word; -webkit-tap-highlight-color: transparent; outline: none;">ENERGY":{Current":[1.320,2.100]}</code></p><div class="md-typeset__scrollwrap" style="box-sizing: inherit; margin: 1em -0.8rem; overflow-x: auto; touch-action: auto; color: rgba(0, 0, 0, 0.87); font-family: Barlow, -apple-system, BlinkMacSystemFont, Helvetica, Arial, sans-serif; font-size: 17.6px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(255, 255, 255); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;"><div class="md-typeset__table" style="box-sizing: inherit; display: inline-block; margin-bottom: 0.5em; padding: 0px 0.8rem;">

Trigger | When it occurs
-- | --
Energy#Current[N] | N = Number of the field. 1 for the first 1.320, 2 for the second 2.100 etc.
Energy#Current[1]>1.000 | whenever the first value of Energy#Current is higher than 1.000.

</div></div>
</body>
</html>


Based on this I expected my trigger to work. Possibly some special code should be enabled for arrays to work?


I also tried the rule that you proposed (although not practical because in reality the zone_dict would contain text, but just to know), but this construction does not result in a trigger. Also the version with Event#SIPupdate.S02 does not trigger.

This version does work:
```
Rule1 
ON mqtt#connected DO Subscribe SIPupdate, SIP/zones, zone_dict.S01 ENDON
ON Event#SIPupdate DO Power2 %value% ENDON

```

**Reply by barbudor** [reply]
I missed that part on array, thanks for pointing it.
I will look further during the week-end

**Reply by barbudor** [reply]
I've made some progress on understanding what is happening.
There are multiple points that make that your initial attempt would not work

1. The array is working only when a rule trigger value is a JSON value like `{"ENERGY":{"Current":[1,2]}}` => `on tele-ENERGY#Current[1] do ...` but as of today EVENT do not support a JSON Payload, only simple string.
So this prevent my suggestion to work.

2.  `Subscribe <event>, <topic>, <key>` extract the value associated to the key as a string. So in your case what is sent to the EVENT is not `{"zone_list": [0, 0, ...]}` but only  `"[0,0, ...]"` which is not a correct JSON and cannot be parsed.

I've already fixed point 1 which allows one of the below 2 possibilities to *almost* works:
```
Subscribe SIPupdate, SIP/zones
Rule1 ON event#SIPupdate#zone_dict#S01 DO .... ENDON
```
Now works because the value passed to the event is the whole valid JSON object `{ ... , "zone_dict": {"S01":0, "S02", ...}` which can now be parsed with the syntax `SIPupdate#zone_dict#S01`
I know this is not what you are looking for ... 😉

But this fix allows the following:
```
Subscribe SIPupdate, SIP/zones
Rule1 ON event#SIPupdate#zone_list[1] DO .... ENDON
```
Which should be close enough to what your are looking for and I would assume that you would be satisfied with that.
What prevents it to work now is that a buffer length that gets overwritten by your long JSON object.


For completeness one thing that still doesn't work is:
```
Subscribe SIPupdate, SIP/zones, zone_dict
Rule1 ON event#SIPupdate#S01 DO .... ENDON
```
Because the extraction done by Subscribe works only if the value of the key is simple and not an object.
Not that you are interested in it, just to say...

Note that there is no way to make your original syntax to work, at least not without breaking a lot. My current patch is simple.

**Comment by CodingEJ** [comment]
That's great progress and even better news!
A little change in syntax to make such triggers working doesn't hurt anyone
I'd say. Your fix would totally solve my problem.
For the buffer size, isn't there a define that could solve this?

Thanks again for your support.


Op za 26 jun. 2021 18:15 schreef Barbudor ***@***.***>:

**Reply by barbudor** [reply]
I made a PR https://github.com/arendst/Tasmota/pull/12496 for the 1st part but the Event name + MQTT payload must remain below 100 chars.
So as of today your MQTT payload that is around 300 bytes is not supported.

For now extending that would require a more changes a few buffers are involved.

**Comment by CodingEJ** [comment]
Is the 100 character limit for the complete mqtt payload or only applicable
to the part that is filtered out to trigger the event?
If the full payload counts I'll definitely have to start thinking about a
different solution because the zone dictionary will be more than 100
characters, even when short descriptions for the zones are used.
I guess I'll setup a translation to send power statements. No problem, just
less elegant.




Op zo 27 jun. 2021 20:20 schreef Barbudor ***@***.***>:

**Reply by barbudor** [reply]
Filtering works only for single values (int, float, string). Not for sub- objectsor arrays.

To do parsing at the rule trigger, you should not enable filtering but pass the whole MQTT payload, which must be below 100 chars.
I will look further if we can override the length problem or alternatively support array index in the filtering

**Comment by CodingEJ** [comment]
All clear now. Thanks.

Op zo 27 jun. 2021 20:46 schreef Barbudor ***@***.***>:

**Reply by barbudor** [reply]
1st part  PR #12496 has been merged and will be available in the [Development binaries](http://ota.tasmota.com/tasmota/) in a couple of hours.
`ON event#eventname#key[1] DO...` will work as far as the whole MQTT JSON payload is less than 90..100 bytes
